modules/storage/src/storage-provider/providers/aliyun/index.ts

```diff
@@ -0,0 +1,210 @@
+    return true;
+  }
+
+  async folderExists(name: string): Promise<boolean | Error> {
+    return await this._ossClient
+      .head(name)
+      .then(() => true)
+      .catch(err => {
+        if (err.status === 404) return false;
+        throw err;
+      });
+  }
+
+  async deleteFolder(name: string): Promise<boolean | Error> {
+    let mark = null;
+
+    do {
+      const res = await this._ossClient.list(
+        {
+          prefix: name,
+          ['max-keys']: 1000,
+        },
+        {},
+      );
+      mark = res.nextMarker;
+
+      await this._ossClient.deleteMulti(res.objects.map(it => it.name));
+    } while (mark);
+
+    return true;
+  }
+
+  async store(
+    fileName: string,
+    data: any,
+    isPublic: boolean = false,
+  ): Promise<boolean | Error> {
+    await this._ossClient.put(fileName, data, {
+      headers: {
+        'x-oss-object-acl': isPublic ? 'public-read' : 'private',
+      },
+    });
+
+    return true;
+  }
+
+  async delete(fileName: string): Promise<boolean | Error> {
+    await this._ossClient.delete(fileName);
+
+    return true;
+  }
+
+  async exists(fileName: string): Promise<boolean | Error> {
+    return await this._ossClient
+      .head(fileName)
+      .then(() => true)
+      .catch(err => {
+        if (err.status === 404) return false;
+        throw err;
+      });
+  }
+
+  async get(fileName: string, downloadPath?: string): Promise<any | Error> {
+    let { content } = await this._ossClient.get(fileName);
+
+    if (downloadPath) {
+      fs.writeFileSync(downloadPath, content);
+    }
+
+    return content;
+  }
+
+  async getSignedUrl(fileName: string): Promise<any | Error> {
+    const url = this._ossClient.signatureUrl(fileName, {
+      expires: 3600,
+      method: 'GET',
+    });
+
+    return url;
+  }
+
+  async getPublicUrl(fileName: string): Promise<any | Error> {
+    const url = this._ossClient.signatureUrl(fileName, {
+      expires: 3600 * 24 * 365 * 100,
+      method: 'GET',
+    });
+
+    return url;
+  }
+
+  async rename(currentFilename: string, newFilename: string): Promise<boolean | Error> {
+    await this._ossClient.copy(newFilename, currentFilename);
+    await this.delete(currentFilename);
+    return true;
+  }
+
+  async moveToFolder(filename: string, newFolder: string): Promise<boolean | Error> {
+    await this._ossClient.copy(newFolder + basename(filename), filename);
+    await this.delete(filename);
+    return true;
+  }
+
+  async moveToFolderAndRename(
+    currentFilename: string,
+    newFilename: string,
+    newFolder: string,
+  ): Promise<boolean | Error> {
+    await this._ossClient.copy(newFolder + newFilename, currentFilename);
+    await this.delete(currentFilename);
+    return true;
+  }
+
+  async moveToContainer(
+    filename: string,
+    newContainer: string,
+  ): Promise<boolean | Error> {
+    const oldBucket = this._activeContainer;
+
+    this._ossClient.useBucket(newContainer);
+
+    // FIXME copy api type error
```

Comment:
Please elaborate on these type casts.
`oldBucket` should be a string, whereas `CopyObjectOptions` is an object of type `{ timeout?: number | undefined; meta?: UserMeta | undefined; headers?: object | undefined; }`

Reply:
This is a @types/ali-oss problem,

in ali-oss sdk, the copy api is `.copy(name, sourceName[, sourceBucket, options])
`, but in @types/ali-oss copy type is `copy(name: string, sourceName: string, options?: OSS.CopyObjectOptions): Promise<OSS.CopyAndPutMetaResult>;`
https://github.com/ali-sdk/ali-oss#copyname-sourcename-sourcebucket-options

The reason I did this was to fool the type check

I have already submit a fix PR to types/ali-oss
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60757

Reply:
Understandable. Workaround should do for now.
I subscribed to the thread. Hopefully we won't need it for long.

Reply:
@kon14   https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60757  has be merged. this code has updated